packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs

```diff
@@ -376,6 +493,333 @@ private MethodBodyStatement GetPropertyInitializers(IReadOnlyList<ParameterProvi
+            out ValueExpression value)
+        {
+            var deserializeDictionaryStatement = new MethodBodyStatement[]
+            {
+                Declare("dictionary", New.Dictionary(valueType.Arguments[0], valueType.Arguments[1]), out var dictionary),
+                new ForeachStatement("prop", jsonElement.EnumerateObject(), out var prop)
+                {
+                   CreateDeserializeDictionaryValueStatement(valueType.ElementType, dictionary, new JsonPropertySnippet(prop), serializationFormat)
+                }
+            };
+            value = dictionary;
+            return deserializeDictionaryStatement;
+        }
+
+        private ValueExpression CreateDeserializeValueExpression(CSharpType valueType, SerializationFormat serializationFormat, JsonElementSnippet jsonElement) =>
+            valueType switch
+            {
+                { SerializeAs: { } serializeAs } =>
+                    new CastExpression(GetValueTypeDeserializationExpression(serializeAs, jsonElement, serializationFormat), valueType),
+                { IsFrameworkType: true } when valueType.FrameworkType == typeof(Nullable<>) =>
+                    GetValueTypeDeserializationExpression(valueType.Arguments[0].FrameworkType, jsonElement, serializationFormat),
+                { IsFrameworkType: true } =>
+                    GetValueTypeDeserializationExpression(valueType.FrameworkType, jsonElement, serializationFormat),
+                _ =>
+                    CreateDeserializeValueExpressionForProvider(valueType.Implementation, jsonElement) ?? throw new InvalidOperationException($"Unable to deserialize type {valueType}")
+            };
+
+        private ValueExpression? CreateDeserializeValueExpressionForProvider(TypeProvider implementation, JsonElementSnippet jsonElement) =>
+            implementation switch
+            {
+                EnumProvider enumProvider =>
+                    enumProvider.ToEnum(GetValueTypeDeserializationExpression(enumProvider.ValueType.FrameworkType, jsonElement, SerializationFormat.Default)),
+                ModelProvider modelProvider =>
+                    new InvokeStaticMethodExpression(modelProvider.Type, $"Deserialize{modelProvider.Name}", [jsonElement, _mrwOptionsDeserializationParameterSnippet]),
+                _ => null
+            };
+
+        private MethodBodyStatement CreateDeserializeArrayItemIntoListStatement(
+            CSharpType arrayItemType,
+            ListSnippet listVar,
+            JsonElementSnippet arrayItemVar,
+            SerializationFormat serializationFormat) =>
+            NullCheckCollectionItemIfRequired(arrayItemType, arrayItemVar, listVar.Add(Null), new MethodBodyStatement[]
+            {
+                DeserializeValue(arrayItemType, arrayItemVar, serializationFormat, out ValueExpression value),
+                listVar.Add(value),
+            });
+
+        private MethodBodyStatement CreateDeserializeArrayItemIntoArrayStatement(
```

Comment:
Lets try to avoid methods that are only called in one place.  Most of the time stack vars can solve this for us and makes the code much easier to follow